File: python/src/bursar/repositories/billing/topup.py
from __future__ import annotations

from bursar.repositories._types import QueryFn
from bursar.repositories._utils import unwrap_jsonb, validate_non_empty
from bursar.repositories.schemas import BillingTopupRow


class BillingTopupRepository:
    """Repository for credit topup resolution operations.

    All methods call Postgres via raw SQL queries through the query function.
    Returns None when the query returns no rows.
    """

    def __init__(self, execute: QueryFn) -> None:
        self._execute = execute

    def resolve_by_price(
        self,
        provider: str,
        price_id: str | None,
        product_id: str | None,
    ) -> BillingTopupRow | None:
        """Resolve a credit topup by provider and price/product IDs.

        Args:
            provider: The billing provider identifier.
            price_id: The provider price ID, or None.
            product_id: The provider product ID, or None.

        Returns:
            BillingTopupRow if found, None otherwise.
        """
        validate_non_empty(provider, "provider")
        row = unwrap_jsonb(
            self._execute(
                "SELECT * FROM public.resolve_credit_topup_by_price(%s, %s, %s)",
                [provider, price_id, product_id],
            )
        )
        return BillingTopupRow.model_validate(row) if row else None

    def resolve_by_lookup(self, provider: str, lookup_key: str) -> BillingTopupRow | None:
        """Resolve a credit topup by provider and lookup key.

        Args:
            provider: The billing provider identifier.
            lookup_key: The topup lookup key.

        Returns:
            BillingTopupRow if found, None otherwise.
        """
        validate_non_empty(provider, "provider")
        row = unwrap_jsonb(
            self._execute(
                "SELECT * FROM public.resolve_credit_topup_by_lookup(%s, %s)",
                [provider, lookup_key],
            )
        )
        return BillingTopupRow.model_validate(row) if row else None

File: javascript/src/repositories/billing/topup.ts
import { z } from "zod";
import type { QueryFn } from "../types.js";
import { unwrapJsonb, safeParse } from "../_shared.js";

export const BillingTopupRowSchema = z
  .object({
    topup_key: z.string().optional(),
    credits_per_unit: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    credits_per_major_unit: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    tier: z.string().optional(),
    deposit_to: z.string().optional(),
  })
  .passthrough();

export type BillingTopupRow = z.infer<typeof BillingTopupRowSchema>;

/** Repository for credit top-up resolution. */
export class BillingTopupRepository {
  constructor(private query: QueryFn) {}

  /** Resolve a credit top-up by price ID and product ID. */
  async resolveByPrice(
    provider: string,
    priceId: string | null,
    productId: string | null,
  ): Promise<BillingTopupRow | null> {
    const rows = await this.query(
      "SELECT * FROM bursar.resolve_credit_topup_by_price($1, $2, $3)",
      [provider, priceId, productId],
    );
    const data = unwrapJsonb(rows);
    return data
      ? safeParse(BillingTopupRowSchema, data, "BillingTopupRepository.resolveByPrice")
      : null;
  }

  /** Resolve a credit top-up by provider lookup key. */
  async resolveByLookup(provider: string, lookupKey: string): Promise<BillingTopupRow | null> {
    const rows = await this.query("SELECT * FROM bursar.resolve_credit_topup_by_lookup($1, $2)", [
      provider,
      lookupKey,
    ]);
    const data = unwrapJsonb(rows);
    return data
      ? safeParse(BillingTopupRowSchema, data, "BillingTopupRepository.resolveByLookup")
      : null;
  }
}
